tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb

```diff
@@ -0,0 +1,502 @@
+    "\\end{equation*}\n",
+    "\n",
+    "\\begin{equation*}\n",
+    "recall = \\frac{tp}{(tp + fn)}\n",
+    "\\end{equation*}\n",
+    "\n",
+    "\\begin{equation*}\n",
+    "F1 = 2 \\cdot \\frac{ (precision \\cdot recall)}{(precision + recall)}\n",
+    "\\end{equation*}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# II. Labeling Functions\n",
+    "\n",
+    "## A. Pattern Matching Labeling Functions\n",
+    "\n",
+    "One powerful form of labeling function design is defining sets of keywords or regular expressions that, as a human labeler, you know are correlated with the true label. For example, we could define a dictionary of terms that occur between person names in a candidate. One simple dictionary of terms indicating a true relation could be:\n",
+    "    \n",
+    "    spouses = {'husband', 'wife'}\n",
+    " \n",
+    "We can then write a labeling function that checks for a match with these terms in the text that occurs between person names. To access the text between the person mentions, we can use the `get_between_tokens` preprocessors from the previous notebook.\n",
+    "\n",
+    "    @labeling_function(resources=dict(spouses=spouses), preprocessors=[get_between_tokens])\n",
+    "    def LF_husband_wife(x: DataPoint, spouses: List[str]) -> int:\n",
+    "        return 1 if len(spouses.intersection(set(x.between_tokens))) > 0 else 0\n",
+    "        \n",
+    "The idea is that we can easily create dictionaries that encode themes or categories descibing all kinds of relationships between 2 people and then use these objects to _weakly supervise_ our classification task.\n",
+    "\n",
+    "    other_relationship = {'boyfriend', 'girlfriend'}\n",
+    "    \n",
+    "**IMPORTANT** Good labeling functions manage a trade-off between high coverage and high precision. When constructing your dictionaries, think about building larger, noiser sets of terms instead of relying on 1 or 2 keywords. Sometimes a single word can be very predictive (e.g., `ex-wife`) but it's almost always better to define something more general, such as a regular expression pattern capturing _any_ string with the `ex-` prefix. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import List\n",
+    "\n",
+    "from snorkel.labeling.apply import PandasLFApplier\n",
+    "from snorkel.labeling.lf import labeling_function\n",
+    "from snorkel.types import DataPoint\n",
+    "\n",
+    "from spouse_preprocessors import get_left_tokens, \\\n",
+    "get_person_last_names, get_person_text\n",
+    "    \n",
+    "\n",
+    "spouses = {'spouse', 'wife', 'husband', 'ex-wife', 'ex-husband'}\n",
+    "@labeling_function(resources=dict(spouses=spouses))\n",
+    "def LF_husband_wife(x: DataPoint, spouses: List[str]) -> int:\n",
+    "    return 1 if len(spouses.intersection(set(x.between_tokens))) > 0 else 0"
```

Comment:
Let's do POSITIVE = 1, ABSTAIN = 0, NEGATIVE = -1 somewhere, then use those for all LFs